### ISSUE TITLE
gen_udp: add_membership socket option does not allow to specify the interface index

### ISSUE THREAD

**Issue by miaxon** [erlang/otp#7736]
The documentation describes the type `ip_membership()`:

```
ip_membership() =
    {MultiAddress :: inet:ip4_address(),
     Interface :: inet:ip4_address()} |
    {MultiAddress :: inet:ip4_address(),
     Address :: inet:ip4_address(),
     IfIndex :: integer()}


The tuple with size 3 is *not* supported on all platforms. 'ifindex' defaults to zero (0) on platforms that supports the 3-tuple variant.
```
I would like the field `IfIndex` to be enabled so that the correct interface number can be specified.

Using the utility `strace`, I looked at what happens when a socket is created and found out this:

```
erlang: 
  gen_udp:open(1234, [{add_membership,{{232,252,0,100},{0,0,0,0},6}}]).

strace:
  setsockopt(18, SOL_IP, IP_ADD_MEMBERSHIP, {imr_multiaddr=inet_addr("232.252.0.100"), imr_interface=inet_addr("0.0.0.0")}, 12) = -1 ENODEV (No such device)
```
Interface at index 6 exists on my system:

```
ip a
....

2: enp7s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 00:e0:8a:68:48:3a brd ff:ff:ff:ff:ff:ff
    inet 192.168.0.12/24 brd 192.168.0.255 scope global dynamic noprefixroute enp7s0
       valid_lft 13875sec preferred_lft 13875sec
    inet6 fd1b:692f:d370:1:5b7d:b2fc:6e5b:6779/64 scope global temporary dynamic 
       valid_lft 568277sec preferred_lft 49583sec
    inet6 fd1b:692f:d370:1:7020:784:b650:8ac9/64 scope global dynamic mngtmpaddr noprefixroute 
       valid_lft 604748sec preferred_lft 86348sec
    inet6 fe80::9775:8571:4666:d153/64 scope link noprefixroute 
       valid_lft forever preferred_lft forever

....

6: veth0@if5: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether ea:ae:e2:5b:c1:0d brd ff:ff:ff:ff:ff:ff link-netns G
    inet 198.51.100.1/30 brd 198.51.100.3 scope global veth0
       valid_lft forever preferred_lft forever
    inet6 fe80::e8ae:e2ff:fe5b:c10d/64 scope link 
       valid_lft forever preferred_lft forever
```

Interface is NOT added to the group and do not receive multicast, another interface is added instead:

```
netstat -g|grep 232.252.0.100
enp7s0          1      232.252.0.100
```
My system is `Linux  6.1.0-13-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.55-1 (2023-09-29) x86_64 GNU/Linux`

In plain C working now

```
struct ip_mreqn mreq;
mreq.imr_multiaddr.s_addr = 0x6400fce8;
mreq.imr_address.s_addr = 0;
mreq.imr_ifindex = 6;
setsockopt(sock, IPPROTO_IP, IP_ADD_MEMBERSHIP, &mreq, sizeof(mreq); // success, no errors
```
The interface is added to the group correctly and receive multicast:

```
netstat -g|grep 232.252.0.100
veth0           1      232.252.0.100
```

This is important in order to accept multicast on a specific interface WITHOUT specifying its IP address.

**Comment by miaxon** [comment]
Please, check PR #7761

**Comment by RaimoNiskanen** [comment]
Fixed in PR #7761.
